src/components/js-tracer/ExpensiveChart.js

```diff
@@ -0,0 +1,167 @@
+|};
+
+type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+
+/**
+ * This component uses the timing information for JS tracer events. This information
+ * can be quite expensive to compute. This component is responsible for encapsulating
+ * the complexity of deferring that calculation until the proper moment in the
+ * component life cycle.
+ */
+class JsTracerExpensiveChart extends React.PureComponent<Props> {
+  /**
+   * Determine the maximum zoom of the viewport.
+   */
+  getMaximumZoom(): UnitIntervalOfProfileRange {
+    const { timeRange: { start, end }, interval } = this.props;
+    return interval / (end - start);
+  }
+
+  render() {
+    const {
+      timeRange,
+      threadIndex,
+      jsTracerTable,
+      previewSelection,
+      updatePreviewSelection,
+      showJsTracerSummary,
+    } = this.props;
+
+    const jsTracerTimingRows = _getJsTracerTimingRows(
+      jsTracerTable,
+      showJsTracerSummary
+    );
+
+    // The viewport needs to know about the height of what it's drawing, calculate
+    // that here at the top level component.
+    const maxViewportHeight = jsTracerTimingRows.length * ROW_HEIGHT;
+
+    return (
+      <JsTracerCanvas
+        key={threadIndex}
+        viewportProps={{
+          timeRange,
+          previewSelection,
+          maxViewportHeight,
+          viewportNeedsUpdate,
+          maximumZoom: this.getMaximumZoom(),
+          marginLeft: TIMELINE_MARGIN_LEFT,
+          marginRight: TIMELINE_MARGIN_RIGHT,
+        }}
+        chartProps={{
+          jsTracerTimingRows,
+          jsTracerTable,
+          updatePreviewSelection,
+          rangeStart: timeRange.start,
+          rangeEnd: timeRange.end,
+          rowHeight: ROW_HEIGHT,
+          threadIndex,
+        }}
+      />
+    );
+  }
+}
+
+type JsTracerTimingWeakmap = WeakMap<JsTracerTable, JsTracerTiming[]>;
+const _jsTracerTimingWeakmap: JsTracerTimingWeakmap = new WeakMap();
+const _jsTracerLeafTimingWeakmap: JsTracerTimingWeakmap = new WeakMap();
+
+/**
+ * Don't compute the JS tracer timing information in the selector, as it's very expensive.
+ * Defer computation until we actually use it in the render. Sometimes the selector
+ * is run, but the component is not actually rendered. Running this code in the selector
+ * would jank the browser, without updating the UI with a "loading" notification.
+ */
+function _getJsTracerTimingRows(
+  jsTracerTable: JsTracerTable,
+  showJsTracerSummary: boolean
+): JsTracerTiming[] {
+  if (showJsTracerSummary) {
+    let timing = _jsTracerTimingWeakmap.get(jsTracerTable);
+    if (!timing) {
+      timing = JsTracer.getJsTracerLeafTiming(jsTracerTable);
+      _jsTracerTimingWeakmap.set(jsTracerTable, timing);
+    }
+    return timing;
+  }
+  let timing = _jsTracerLeafTimingWeakmap.get(jsTracerTable);
+  if (!timing) {
+    timing = JsTracer.getJsTracerTiming(jsTracerTable);
+    _jsTracerLeafTimingWeakmap.set(jsTracerTable, timing);
+  }
```

Comment:
you used `_jsTracerLeafTimingWeakmap` with `getJsTracerTiming` and `_jsTracerTimingWeakmap` with `getJsTracerLeafTiming`. This still works but this is confusing to read :-)

Reply:
Woops, resolved by changing to a keyed component.